### DISCUSSION TITLE
Did a thing. HammerPost

### DISCUSSION THREAD

**Discussion by dineshtessell** [TPC-Council/HammerDB#527] [Show and tell]
We have created a wrapper around HammerDB to enable PostgreSQL parameter profiling. We would like to express our gratitude towards HammerDB developers and Steve Shaw, which is an amazing tool for simulating live e-commerce workloads and providing results such as `TPM` and `NOPM`. We use this workload to collect node metrics for each parameter set on PostgreSQL and MySQL.

This tool is incredibly helpful in identifying the best parameter profile for a given workload. I hope that it will be useful for others who want to fine-tune their PostgreSQL parameters.

Looking forward for your inputs or suggestions on this [tool](https://github.com/dineshkumar02/hammerpost).

https://github.com/dineshkumar02/hammerpost

Thank you again.

Dinesh Kumar

**Comment by sm-shaw** [comment]
Hi, 

This looks very interesting and a great addition for people running HammerDB on PostgreSQL in an automated environment. 
I am trying it out, but I am at the step as follows: 

**> Please make sure, you started the hammerpost-agent serivce on PostgreSQL node, before performing below steps.**

Can you describe how to run the hammerpost-agent service? I can't seem to find the instructions to do that, unless I missed it? 

Also interesting to see how big the Go executables is. Is there any way to optimize this? Its not essential but its a lot bigger than everything else. 

Steve

**Reply by dineshtessell** [reply]
Hi @sm-shaw ,

Thank you for your kind words.

I am so sorry, seems I missed to put the instructions into the doc about [hammerpost-agent](https://github.com/dineshkumar02/hammerpost-agent). I will fix it.

Here is the command, which we used to prepare this demo.

`./hammerpost-agent --stop-cmd "/usr/pgsql-15/bin/pg_ctl -D /var/lib/pgsql/15/data stop -mf" --start-cmd "/usr/pgsql-15/bin/pg_ctl -D /var/lib/pgsql/15/data start -l /tmp/startup.log" --pgdsn "postgres://postgres:postgres@localhost:5432/postgres" --db-type postgres
`

HammerPost also works with MySQL(Did couple of tests, seems things are pretty stable with MySQL too).

**> Also interesting to see how big the Go executables is. Is there any way to optimize this? Its not essential but its a lot bigger than everything else.**
Good suggestion, we will double check which dependency is causing this much of binary size.

**Comment by sm-shaw** [comment]
Hi, 
Many thanks, yes, that was it.  I needed to do the same build on [hammerpost-agent](https://github.com/dineshkumar02/hammerpost-agent) and copy it into the HammerDB directory, then it works. 
This looks excellent and with a lot of potential.  Also, if it works against MySQL then it should work against MariaDB as well I would think? 
I have tried PostgreSQL. I needed to add a postgresql role for everything to work. Sometimes I have found I have need to start the database manually between tests, when I see `error in applying above parameters: failed to connect` as in the screenshot:
however, I am running on virtual box on an older PC, so it could be that this is likely that the shutdown is taking longer than expected and would work on a server where it is intended.  
The agent shows:
`stopping database using command:  /home/hammerdb/pgsql/bin/pg_ctl -D /home/hammerdb/pgsql/data stop -mf`
An example of it working is here for my tests:
![hammerpost](https://user-images.githubusercontent.com/38044085/227193514-c49f8444-0b15-4901-b803-0361085afc34.PNG)
One thing that may be useful is that there are HammerDB images on Docker Hub with the main one including all databases and lighter weight ones per database. These include the necessary drivers already, so you wouldn't need the first steps of installing the driver and doing librarycheck as everything would already be there.  (There could even be the possibility of building a container with everything)
For the Go executable size I think this is a feature of Go so shouldn't be a problem as they are not too big. 
More documentation and examples would definitely help people work out the steps to getting started, as this would be really useful in testing out different parameters. 
Great Work! :smiley:

**Reply by dineshtessell** [reply]
Hi @sm-shaw ,

Thank you for your input and suggestions. They are very encouraging to us.
I don't believe that you tried "hammerpost" in your busy schedule. Thank you :-)

Yes, this tool should work with MariaDB and MySQL too.
MariaDB I didn't tried, but hoping it will work.

We will fix the documentation and provide more examples.

- -Dinesh Kumar

**Reply by sm-shaw** [reply]
@dineshtessell, no problem at all, with open source contributions are always the most important thing so I will always make time. 
Yes if it works againt MySQL it should work against MariaDB with very little change, if any. In fact it would be an excellent tool to compare them for modifiying the same parameters on each. 
So more documentation and examples would be good, especially on viewing the additional statistics you gather. 
When I get the chance I will aim to add a blog post to the HammerDB site to point to your work.

**Reply by dineshtessell** [reply]
@sm-shaw Thank you so much for your inputs.
Improved README file, and also included necessary examples for the users.
Hope it is helpful to start performing the benchmarks on PostgreSQL/MySQL and MariaDB with HammerDB.

Thank You 🙏

**Reply by sm-shaw** [reply]
This looks good.  The only change I would suggest is to modify the docker url to https://hub.docker.com/r/tpcorg/hammerdb this is the official image from the TPC (tpc.org) so this is the one that is up-to-date with all the latest changes.

**Reply by dineshtessell** [reply]
Sure @sm-shaw , will do it

**Reply by sm-shaw** [reply]
I have put a blog post here https://www.hammerdb.com/blog/uncategorized/find-the-right-postgresql-mysql-and-mariadb-parameters-first-time-with-hammerpost/ to encourage people to take a look at the project.

I installed hammerpost in the docker container, I also found as a correction to my previous comments that starting the agent with su previleges resolved the connection issues and everything ran smoothly. This may be something to do with the container, however the example all worked perfectly for me.

**Reply by dineshtessell** [reply]
Thank you so much @sm-shaw
